units/generate_tasks.go

```diff
@@ -113,93 +123,43 @@ func (j *generateTasksJob) generate(ctx context.Context, t *task.Task) error {
-		"message":       "generate.tasks timing",
-		"function":      "generate",
-		"operation":     "parseProjects",
-		"duration_secs": time.Since(start).Seconds(),
-		"task":          t.Id,
-		"job":           j.ID(),
-		"version":       t.Version,
-	})
-	start = time.Now()
 
-	g, err := model.MergeGeneratedProjects(projects)
-	grip.Debug(message.Fields{
-		"message":       "generate.tasks timing",
-		"function":      "generate",
-		"operation":     "MergeGeneratedProjects",
-		"duration_secs": time.Since(start).Seconds(),
-		"task":          t.Id,
-		"job":           j.ID(),
-		"version":       t.Version,
-	})
+	g, err := model.MergeGeneratedProjects(ctx, projects)
 	if err != nil {
 		return errors.Wrap(err, "merging generated projects")
 	}
 	g.Task = t
 
-	start = time.Now()
-	p, pp, v, err := g.NewVersion(project, parserProject, v)
+	p, pp, v, err := g.NewVersion(ctx, project, parserProject, v)
 	if err != nil {
 		return j.handleError(pp, v, errors.WithStack(err))
 	}
-	grip.Debug(message.Fields{
-		"message":       "generate.tasks timing",
-		"function":      "generate",
-		"operation":     "NewVersion",
-		"duration_secs": time.Since(start).Seconds(),
-		"task":          t.Id,
-		"job":           j.ID(),
-		"version":       t.Version,
-	})
+
 	pref, err := model.FindMergedProjectRef(t.Project, t.Version, true)
 	if err != nil {
 		return j.handleError(pp, v, errors.WithStack(err))
 	}
 	if pref == nil {
 		return j.handleError(pp, v, errors.Errorf("project '%s' not found", t.Project))
 	}
-	start = time.Now()
-	if err = validator.CheckProjectConfigurationIsValid(j.env.Settings(), p, pref); err != nil {
+
+	if err = validator.CheckProjectConfigurationIsValid(ctx, j.env.Settings(), p, pref); err != nil {
 		return j.handleError(pp, v, errors.WithStack(err))
 	}
-	grip.Debug(message.Fields{
-		"message":       "generate.tasks timing",
-		"function":      "generate",
-		"operation":     "CheckProjectConfigurationIsValid",
-		"duration_secs": time.Since(start).Seconds(),
-		"task":          t.Id,
-		"job":           j.ID(),
-		"version":       t.Version,
-	})
 
-	start = time.Now()
-	if err := g.CheckForCycles(v, p, pref); err != nil {
+	if err := g.CheckForCycles(ctx, v, p, pref); err != nil {
 		return errors.Wrap(err, "checking new dependency graph for cycles")
 	}
 
-	grip.Debug(message.Fields{
-		"message":       "generate.tasks timing",
-		"function":      "generate",
-		"operation":     "SimulateNewDependencyGraph",
-		"duration_secs": time.Since(start).Seconds(),
-		"task":          t.Id,
-		"job":           j.ID(),
-		"version":       t.Version,
-	})
-
-	start = time.Now()
-
 	// Don't use the job's context, because it's better to try finishing than to
 	// exit early after a SIGTERM from app server shutdown.
 	saveCtx, saveCancel := context.WithTimeout(context.Background(), 30*time.Minute)
 	defer saveCancel()
-	err = g.Save(saveCtx, j.env.Settings(), p, pp, v)
+	err = g.Save(ctx, saveCtx, j.env.Settings(), p, pp, v)
```

Comment:
an alternative to passing through the job context could be to store the `SpanContext` in the `saveCtx`.
This could look something like
```go
saveCtx, saveCancel := context.WithTimeout(context.Background(), 30*time.Minute)
defer saveCancel()

// Inject the span context into the saveCtx.
saveCtx = trace.ContextWithSpanContext(saveCtx, span.SpanContext())
err = g.Save(ctx, saveCtx, j.env.Settings(), p, pp, v)
...
```

Reply:
Ah that is much cleaner thanks